File: samplecode/thread/app/Thread.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License..

#include <thread>
#include <stdio.h>
using namespace std;

#include "App.h"
#include "Enclave_u.h"

void data_producer(void)
{
    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
    ret = ecall_producer(global_eid);
    if (ret != SGX_SUCCESS)
        abort();
}

void data_consumer(void)
{
    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
    ret = ecall_consumer(global_eid);
    if (ret != SGX_SUCCESS)
        abort();
}

void data_init(void)
{
    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
    ret = ecall_initialize(global_eid);
    if (ret != SGX_SUCCESS)
        abort();
}

void data_uninit(void)
{
    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
    ret = ecall_uninitialize(global_eid);
    if (ret != SGX_SUCCESS)
        abort();
}

/* ecall_thread_functions:
 *   Invokes thread functions including mutex, condition variable, etc.
 */
void ecall_thread_functions(void)
{
    data_init();

    printf("Info: executing thread synchronization, please wait...  \n");
    /* condition variable */
    thread consumer1(data_consumer);
    thread producer0(data_producer);
    thread producer1(data_producer);
    thread producer2(data_producer);
    thread producer3(data_producer);
    thread consumer2(data_consumer);
    thread consumer3(data_consumer);
    thread consumer4(data_consumer);

    consumer1.join();
    consumer2.join();
    consumer3.join();
    consumer4.join();
    producer0.join();
    producer1.join();
    producer2.join();
    producer3.join();

    printf("Info: thread finish...  \n");

    data_uninit();
}

File: sgx_oc/src/linux/ocall/thread.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use super::*;

pub unsafe fn sched_yield() -> OCallResult<()> {
    let mut result: c_int = 0;
    let mut error: c_int = 0;

    let status = u_sched_yield_ocall(&mut result as *mut c_int, &mut error as *mut c_int);

    ensure!(status.is_success(), esgx!(status));
    ensure!(result == 0, eos!(error));
    Ok(())
}

pub unsafe fn nanosleep(req: &mut timespec) -> OCallResult<()> {
    let mut result: c_int = 0;
    let mut error: c_int = 0;

    let mut rem = *req;
    let status = u_nanosleep_ocall(
        &mut result as *mut c_int,
        &mut error as *mut c_int,
        req as *const timespec,
        &mut rem as *mut timespec,
    );

    ensure!(status.is_success(), esgx!(status));
    ensure!(rem.tv_sec <= req.tv_sec, ecust!("Malformed remaining time"));
    if rem.tv_sec == req.tv_sec {
        ensure!(
            rem.tv_nsec < req.tv_nsec,
            ecust!("Malformed remaining time")
        );
    }

    req.tv_sec = rem.tv_sec;
    req.tv_nsec = rem.tv_nsec;

    ensure!(result == 0, eos!(error));

    req.tv_sec = 0;
    req.tv_nsec = 0;
    Ok(())
}
